### PR TITLE
CCIP-4659 New Modified ERC165Checker library

### PR DESCRIPTION
**Pull request by jhweintraub**
To prevent issues with the OZ ERC165Checker's enforcement of 32k gas on external calls, this library adds a check that enough gas exists before making the call, and reverts if it does not. 

The OffRamp will use this version of the supportsInterface library to prevent message delivery issues. 
The 31k gas requirement was chosen due to the 63/64 gas rule and that the staticcall required 30k gas. 

```
30k gas is required to make the staticcall. Under the 63/64 rule this means that 30,477 gas must be available
  /// to ensure that at least 30k is forwarded. Checking for at least 31,000 ensures that after additional
  /// operations are performed there is still >= 30,477 gas remaining.
  /// 30,000 = ((30,477 * 63) / 64)
  ```

### CODE REVIEW

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * @dev Library used to query support of an interface declared via {IERC165}.
+ *
+ * Note that these functions return the actual result of the query: they do not
+ * `revert` if an interface is not supported. It is up to the caller to decide
+ * what to do in these cases.
+ *
+ * Note this is exactly the same as the OZ version, with the exception that external calls will revert
+ * if < 31_000 gas is available, to prevent message delivery issues in CCIP.
+ */
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: let's fully convert it to our style, including the comments using `///`

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 2]
```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: nit: rm newline

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ * Note that these functions return the actual result of the query: they do not
+ * `revert` if an interface is not supported. It is up to the caller to decide
+ * what to do in these cases.
+ *
+ * Note this is exactly the same as the OZ version, with the exception that external calls will revert
+ * if < 31_000 gas is available, to prevent message delivery issues in CCIP.
+ */
+library ERC165CheckerReverting {
+  // As per the EIP-165 spec, no interface should ever match 0xffffffff
+  bytes4 private constant _INTERFACE_ID_INVALID = 0xffffffff;
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: I don't think we need an underscore here?

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  // bytes4(keccak256("NotEnoughGasForSupportsInterfaceCall()"))
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: probably good to also define the error so we can reference it in tests and check it's actually the same

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  /**
+   * @dev Returns true if `account` supports all the interfaces defined in
+   * `interfaceIds`. Support for {IERC165} itself is queried automatically.
+   *
+   * Batch-querying can lead to gas savings by skipping repeated checks for
+   * {IERC165} support.
+   *
+   * See {IERC165-supportsInterface}.
+   */
+  function _supportsAllInterfacesReverting(address account, bytes4[] memory interfaceIds) internal view returns (bool) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: can probably keep this version minimal and only include things we're actually using

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * @dev Library used to query support of an interface declared via {IERC165}.
+ *
+ * Note that these functions return the actual result of the query: they do not
+ * `revert` if an interface is not supported. It is up to the caller to decide
+ * what to do in these cases.
+ *
+ * Note this is exactly the same as the OZ version, with the exception that external calls will revert
+ * if < 31_000 gas is available, to prevent message delivery issues in CCIP.
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: Code says 32k

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    uint256 returnSize;
+    uint256 returnValue;
+
+    assembly {
+      // Enforce that there's enough gas avilable so that the call will not fail due to OOG error. Without this supportsInterface() may return false when it should return true.
+
+      // 32,000 gas was chosen to ensure enough gas to invoke the staticcall after the check
+      // without breaking the 63/64 rule. Under EIP-150 there must be at least ~30,476 gas
+      // remaining before the staticcall.
+      if lt(gas(), 32000) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: Let's make this a constant

File: contracts/src/v0.8/ccip/offRamp/OffRamp.sol [hunk 8]
```diff
@@ -604,9 +604,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
     // 3. If the receiver is a contract but it does not support the IAny2EVMMessageReceiver interface.
     //
     // The ordering of these checks is important, as the first check is the cheapest to execute.
+    //
+    // To prevent message delivery bypass issues, a modified version of the ERC165Checker is used exclusively here
```
Comment (RensR) [inline contracts/src/v0.8/ccip/offRamp/OffRamp.sol]: Comment claims exclusive usage but it's also used in the token pool call. Probably not relevant to mention if something is exclusively used here

**Review by RensR — COMMENTED** [review]
Can you show how you got to 32k (or 31k as the comment suggests)?

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol (hunk 3 above)
Comment (jhweintraub) [reply contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: Copied directly from the OZ library. Will remove.

File: contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol (hunk 4 above)
Comment (jhweintraub) [reply contracts/src/v0.8/ccip/libraries/ERC165CheckerReverting.sol]: I'll just replace it with the actual error and then use top-level solidity to insert it into the assembly. Much easier.

File: .github/scripts/map-affected-files-to-modules.sh [hunk 9]
```diff
@@ -21,30 +21,40 @@ echo "Found modules: $modules"
 declare -A unique_modules
 
 for path_to_file in $changed_files; do
-	echo "Resolving a module affected by a file: '$path_to_file'"
```
Comment (RensR) [inline .github/scripts/map-affected-files-to-modules.sh:24]: ???

Comment (RensR) [inline .github/scripts/map-affected-files-to-modules.sh]: ???

Comment (jhweintraub) [reply .github/scripts/map-affected-files-to-modules.sh]: idk why that file got changed but it's fixed now.

**Review by RensR — DISMISSED** [review]

**Review by RensR — APPROVED** [review]

**Review by 0xsuryansh — APPROVED** [review]

**Pull request merged**